File: 字符串/14.最长公共前缀.ts
// 14 最长公共前缀

/*
 * 编写一个函数来查找字符串数组中的最长公共前缀
 * 如果不存在公共前缀，返回空字符串 ""
 */

export function longestCommonPrefix(strs: string[]): string {
  // 先找出最短的字符串
  let min: string = strs[0];
  for (let i = 1; i < strs.length; i++) {
    if (min.length > strs[i].length) {
      min = strs[i];
    }
  }

  let len: number = min.length;
  // 外层循环 min 的长度
  for (let i = 0; i < len; i++) {
    let flag = true;
    for (let j = 0; j < strs.length; j++) {
      if (!strs[j].includes(min)) {
        flag = false;
        continue;
      }
    }
    // 如果不包含，则 min 长度减一
    if (flag === false) {
      min = min.slice(0, min.length - 2);
    } else {
      return min;
    }
  }

  return '';
}

// test
let res1 = longestCommonPrefix(['flower', 'flow', 'flight']);
let res2 = longestCommonPrefix(['dog', 'racecar', 'car']);
let res3 = longestCommonPrefix(['reflower', 'flow', 'flight']);
console.log(res1);
console.log(res2);
console.log(res3);

File: Data Structure & Algorithm/String/zh/14.最长公共前缀/14.最长公共前缀.py
# 14 最长公共前缀

# 编写一个函数来查找字符串数组中的最长公共前缀
# 如果不存在公共前缀，返回空字符串 ""

# 示例 1：
# 输入：strs = ["flower","flow","flight"]
# 输出："fl"

# 示例 2：
# 输入：strs = ["dog","racecar","car"]
# 输出：""
# 解释：输入不存在公共前缀。

from typing import List


class Solution:

    def longestCommonPrefix(self, strs: List[str]) -> str:
        minStr = strs[0]
        length = len(strs)
        for i in range(1, length):
            if len(minStr) > len(strs[i]):
                minStr = strs[i]

        res = ''
        for i in range(len(minStr)):
            flag = True
            for j in range(length):
                if strs[j][i] != minStr[i]:
                    flag = False

            if flag == True:
                res = res + minStr[i]
            else:
                return res

        return res


# test
strs = ["flower", "flow", "flight"]
solution = Solution()
res = solution.longestCommonPrefix(strs)
print(res)
